components/nexus-repository/src/test/java/org/sonatype/nexus/repository/storage/StorageFacetImplIT.java

```diff
@@ -561,4 +604,160 @@ public void duplicateAssetName() throws Exception {
   private void checkSize(Iterable iterable, int expectedSize) {
     assertThat(Iterators.size(iterable.iterator()), is(expectedSize));
   }
+
+  @Test
+  public void repeatedAssetModificationsAreSaved() throws Exception {
+    createComponent("testGroup", "testName", "testVersion");
+    try (StorageTx tx = underTest.openTx()) {
+      final Component component = tx.findComponentWithProperty("version", "testVersion", tx.getBucket());
+      final Asset asset = tx.createAsset(tx.getBucket(), component).name("asset");
+
+      final NestedAttributesMap attributes = asset.formatAttributes();
+      attributes.set("attribute1", "original");
+      tx.saveAsset(asset);
+
+      final Iterable<Asset> assets = tx.browseAssets(component);
+      final Asset reloadedAsset = assets.iterator().next();
+
+      reloadedAsset.formatAttributes().set("attribute2", "alternate");
+      tx.saveAsset(reloadedAsset);
+
+      tx.commit();
+    }
+
+    try (StorageTx tx = underTest.openTx()) {
+      final Component component = tx.findComponentWithProperty("version", "testVersion", tx.getBucket());
+      final Iterable<Asset> assets = tx.browseAssets(component);
+      final Asset asset = assets.iterator().next();
+
+      assertThat(asset.formatAttributes().get("attribute1", String.class), equalTo("original"));
+      assertThat(asset.formatAttributes().get("attribute1", String.class), equalTo("original"));
+    }
+  }
+
+  @Test
+  public void transactionsRollBackWhenRequired() throws Exception {
+    try (StorageTx tx = underTest.openTx()) {
+      final Component component = tx.createComponent(tx.getBucket(), testFormat)
+          .group("myGroup")
+          .version("0.9")
+          .name("myComponent");
+      tx.saveComponent(component);
+      tx.rollback();
+    }
+
+    try (StorageTx tx = underTest.openTx()) {
+      final Component component = tx.findComponentWithProperty("group", "myGroup", tx.getBucket());
+      assertThat(component, is(nullValue()));
+    }
+  }
+
+  @Test
+  public void transactionsRollBackOnException() throws Exception {
+    try {
+      try (StorageTx tx = underTest.openTx()) {
+        final Component component = tx.createComponent(tx.getBucket(), testFormat)
+            .group("myGroup")
+            .version("0.9")
+            .name("myComponent");
+        tx.saveComponent(component);
+        if (true) {
+          throw new MidTransactionException();
+        }
+      }
+    }
+    catch (MidTransactionException ignored) {
+    }
+
+    try (StorageTx tx = underTest.openTx()) {
+      final Component component = tx.findComponentWithProperty("group", "myGroup", tx.getBucket());
```

Comment:
I think this test is a bit bogus, we get the same result regardless of the thrown exception...